File: examples/kafka-jvm/topologies/pixelcount.clj
(ns pixelcount
  (:use [backtype.storm.clojure]
        [pixelcount.spouts.pixel_spout :only [spout] :rename {spout pixel-spout}]
        [streamparse.specs])
  (:gen-class))


(defn pixelcount [options]
   [
    ;; spout configurations
    {"pixel-spout" (spout-spec pixel-spout :p 1)}

    ;; bolt configurations
    {
      ;; Technically, this bolt isn't really needed, just need to add a proper
      ;; deserializer to Kafka spout so that it doesn't stupidly treat all
      ;; messages as strings, but this is fine for a demo
      "pixel-deserializer-bolt" (python-bolt-spec
        options
        {"pixel-spout" :shuffle}
        "bolts.pixel_deserializer.PixelDeserializerBolt"
        ["ip" "ts" "url"]
        :p 1)

      "pixel-count-bolt" (python-bolt-spec
        options
        ;; fields grouping on url
        {"pixel-deserializer-bolt" ["url"]}
        "bolts.pixel_count.PixelCounterBolt"
        ;; terminal bolt
        []
        :p 1
        :conf {"topology.tick.tuple.freq.secs", 1})
    }
  ]
)

File: examples/kafka-jvm/topologies/pixelcount.py
"""
Pixel count topology
"""

from streamparse import Grouping, JavaSpout, Topology

from bolts.pixel_count import PixelCounterBolt
from bolts.pixel_deserializer import PixelDeserializerBolt


class PixelCount(Topology):
    pixel_spout = JavaSpout.spec(
        name="pixel-spout",
        full_class_name="pixelcount.spouts.PixelSpout",
        args_list=[],
        outputs=["pixel"],
    )
    pixel_deserializer = PixelDeserializerBolt.spec(
        name="pixel-deserializer-bolt", inputs=[pixel_spout]
    )
    pixel_counter = PixelCounterBolt.spec(
        name="pixel-count-bolt",
        inputs={pixel_deserializer: Grouping.fields("url")},
        config={"topology.tick.tuple.freq.secs": 1},
    )
